tools/samtools.wdl

```diff
@@ -1006,3 +1018,224 @@ task fixmate {
+    }
+
+    input {
+        File bam
+        String prefix = basename(bam, ".bam") + ".fixmate"
+        Boolean fast_mode = false
+        Boolean add_cigar = true
+        Boolean add_mate_score = true
+        Boolean disable_flag_sanitization = false
+        Boolean disable_proper_pair_check = false
+        Boolean remove_unaligned_and_secondary = false
+        Boolean use_all_cores = false
+        Int ncpu = 2
+        Int modify_memory_gb = 0
+        Int modify_disk_size_gb = 0
+    }
+
+    Float bam_size = size(bam, "GiB")
+    Int memory_gb = ceil(bam_size * 0.2) + 4 + modify_memory_gb
+    Int disk_size_gb = ceil(bam_size * 2) + 10 + modify_disk_size_gb
+
+    command <<<
+        set -euo pipefail
+
+        n_cores=~{ncpu}
+        if ~{use_all_cores}; then
+            n_cores=$(nproc)
+        fi
+        # -1 because samtools uses one more core than `--threads` specifies
+        let "n_cores -= 1"
+
+        samtools collate \
+            --threads "$n_cores" \
+            ~{if fast_mode then "-f" else ""} \
+            -u \
+            -O \
+            ~{bam} \
+            | samtools fixmate \
+                --threads "$n_cores" \
+                -u \
+                ~{if remove_unaligned_and_secondary then "-r" else ""} \
+                ~{if disable_proper_pair_check then "-p" else ""} \
+                ~{if add_cigar then "-c" else ""} \
+                ~{if add_mate_score then "-m" else ""} \
+                ~{if disable_flag_sanitization then "-z off" else ""} \
+                - \
+                - \
+                | samtools sort \
+                    --threads "$n_cores" \
+                    -o ~{prefix + ".bam"} \
+                    -
+    >>>
+
+    output {
+        File fixmate_bam = "~{prefix}.bam"
+    }
+
+    runtime {
+        cpu: ncpu
+        memory: "~{memory_gb} GB"
+        disk: "~{disk_size_gb} GB"
+        container: 'quay.io/biocontainers/samtools:1.19.2--h50ea8bc_0'
+        maxRetries: 1
+    }
+}
+
+task markdup {
+    meta {
+        description: "**[DEPRECATED]** Runs `samtools markdup` on the position-sorted input BAM file. This creates a report and optionally a new BAM with duplicate reads marked."
+        help: "This task assumes `samtools fixmate` has already been run on the input BAM. If it has not, then the output may be incorrect. Running `fixmate` may not be necessary, if the aligner used correctly sets the mate information. This includes but is not limited to having correct insert sizes, mate coordinates, `ms` or \"mate score\" tags, and `MC` or \"mate cigar\" tags. A name-sorted or collated BAM can be run through the `fixmate` task (and then position sorted prior to this task) or a position-sorted BAM can be run through the `position_sorted_fixmate` task. Deprecated due to extremely high memory usage for certain RNA-Seq samples when searching for optical duplicates. Use `mark_duplicates` in `./picard.wdl` instead."
```

Comment:
```suggestion
        help: "This task assumes `samtools fixmate` has already been run on the input BAM. If it has not, then the output may be incorrect. Running `fixmate` may not be necessary, if the aligner correctly sets the mate information. This includes but is not limited to having correct insert sizes, mate coordinates, `ms` or \"mate score\" tags, and `MC` or \"mate cigar\" tags. A name-sorted or collated BAM can be run through the `fixmate` task (and then position-sorted prior to this task) or a position-sorted BAM can be run through the `position_sorted_fixmate` task. Deprecated due to extremely high memory usage for certain RNA-Seq samples when searching for optical duplicates. Use `mark_duplicates` in `./picard.wdl` instead."
```

Reply:
Thank you for bringing my attention back to this. One part here is incorrect:
> Running `fixmate` may not be necessary, if the aligner correctly sets the mate information. This includes but is not limited to having correct insert sizes, mate coordinates, `ms` or \"mate score\" tags,...

No aligner is going to set the `ms` tag. It's not an officially supported tag, it's custom for `fixmate+markdup`. So I'll remove those 2 sentences. `fixmate` _must_ be run. Which is a real pain and downside of the samtools implementation. But that's a tangent.